### PR TITLE
(proof-new) Use bound variable manager

### PR DESCRIPTION
**Pull request by ajreynol**
This uses BoundVarManager for several key places where bound variables are introduced, including for array extensionality witness terms, string preprocessing, quantifiers rewriting, and skolemization.

This is motivated by making certain steps in solving deterministic for the sake of proofs, and gives a more consistent pattern in general for constructing bound variables.

### CODE REVIEW

**Comment by HanielB** [comment]
If I understand correctly this bound variable manager also simplifies stuff unrelated to proofs, like these `getOrMakeBV` whatever functions. It'd be good to add this to the description of the PR.

File: src/theory/quantifiers/quantifiers_rewriter.cpp [hunk 1]
```diff
@@ -987,9 +966,12 @@ bool QuantifiersRewriter::getVarElim(Node n,
...
   if (nk == NOT)
   {
-    return getVarElim(n[0], !pol, args, vars, subs);
+    n = n[0];
+    pol = !pol;
+    nk = n.getKind();
+    Assert(nk != NOT);
   }
-  else if ((nk == AND && pol) || (nk == OR && !pol))
+  if ((nk == AND && pol) || (nk == OR && !pol))
```
Comment (HanielB) [inline src/theory/quantifiers/quantifiers_rewriter.cpp:974]: `else if`?

**Review by HanielB — APPROVED** [review]
LGTM, just a minor comment

File: src/theory/quantifiers/quantifiers_rewriter.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/quantifiers_rewriter.cpp:974]: Actually no, I removed the recursion from the NOT case, so now it should fall through and trigger this.

Comment (HanielB) [reply src/theory/quantifiers/quantifiers_rewriter.cpp:974]: Ah, I did not notice that `nk` is updated.

**Pull request merged**